cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -895,6 +895,84 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_ebb(pos.func, done);
 }
 
+/// Expand shuffle instructions
+fn expand_shuffle(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Shuffle { args, mask, .. } = pos.func.dfg[inst] {
+        // remove the original shuffle instruction
+        pos.remove_inst();
+
+        // extract the original result value and use it (e.g. with_result) as the final result value of our replaced sequences
+        let original_results = pos.func.dfg.detach_results(inst);
+        let original_result = original_results.get(0, &pos.func.dfg.value_lists).unwrap();
+
+        // a mask-building helper: in 128-bit SIMD, 0-15 indicate which lane to read from and a 1 in the most significant position zeroes the lane
+        let build_shuffle_mask = |b: u8| if b > 15 { 0b10000000 } else { b };
+
+        // we only have to worry about aliasing here because copies will be introduced later (in regalloc)
+        let a = pos.func.dfg.resolve_aliases(args[0]);
+        let b = pos.func.dfg.resolve_aliases(args[1]);
+        let mask = pos
+            .func
+            .dfg
+            .immediates
+            .get(mask)
+            .expect("The shuffle immediate should have been recorded before this point")
+            .clone();
+        if a == b {
+            // PSHUFB the first argument (since it is the same as the second)
+            let constructed_mask = mask
+                .iter()
+                .cloned()
+                .map(|b| if b > 15 { b.wrapping_sub(16) } else { b }) // if the mask is greater than 15 it still may be referring to a lane in b
+                .map(build_shuffle_mask)
+                .collect();
+            let handle = pos.func.dfg.constants.insert(constructed_mask);
+            let a_type = pos.func.dfg.value_type(a);
```

Comment:
Rust gets confused:
```
error[E0502]: cannot borrow `pos.func.dfg` as immutable because it is also borrowed as mutable
   --> cranelift-codegen/src/isa/x86/enc_tables.rs:961:47
    |
961 |             let mask_value = pos.ins().vconst(pos.func.dfg.value_type(b).clone(), handle);
    |                              ---       ------ ^^^^^^^^^^^^ immutable borrow occurs here
    |                              |         |
    |                              |         mutable borrow later used by call
    |                              mutable borrow occurs here
```